File: servers/common/payload.go
package common

import (
	"encoding/json"
	"fmt"
)

// Pre-computed JSON payloads for response-size benchmarks. Built once at
// package init so every adapter (and every test) observes byte-identical
// responses on /json-1k and /json-64k. The exact byte length is what
// loadgen records as ThroughputBPS, so any drift here would corrupt the
// throughput axis of the report.
var (
	json1KPayload  []byte
	json64KPayload []byte
)

func init() {
	json1KPayload = generateJSONPayload(1024)
	json64KPayload = generateJSONPayload(65536)
}

// paginatedItem is one row in the paginated response. Field shape and
// order match a realistic CRUD API so frameworks that special-case
// lower-cased keys, RFC 3339 timestamps, etc. do not get a free win.
type paginatedItem struct {
	ID        int    `json:"id"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	Status    string `json:"status"`
	CreatedAt string `json:"created_at"`
}

// paginatedResponse wraps a page of items with pagination metadata.
type paginatedResponse struct {
	Page       int              `json:"page"`
	PerPage    int              `json:"per_page"`
	Total      int              `json:"total"`
	TotalPages int              `json:"total_pages"`
	Data       []*paginatedItem `json:"data"`
}

// generateJSONPayload builds a paginated response of at least targetSize
// bytes by appending items until the marshalled length crosses the
// threshold. The result is deterministic given a fixed targetSize.
func generateJSONPayload(targetSize int) []byte {
	resp := &paginatedResponse{
		Page:       1,
		PerPage:    50,
		Total:      1000,
		TotalPages: 20,
	}
	for i := 1; ; i++ {
		resp.Data = append(resp.Data, &paginatedItem{
			ID:        i,
			Name:      fmt.Sprintf("User %d", i),
			Email:     fmt.Sprintf("user%d@example.com", i),
			Status:    "active",
			CreatedAt: "2024-01-15T09:30:00Z",
		})
		data, _ := json.Marshal(resp)
		if len(data) >= targetSize {
			return data
		}
	}
}

// JSON1KPayload returns the pre-computed ~1 KiB JSON payload. The slice
// is shared — callers must not mutate it.
func JSON1KPayload() []byte { return json1KPayload }

// JSON64KPayload returns the pre-computed ~64 KiB JSON payload. The
// slice is shared — callers must not mutate it.
func JSON64KPayload() []byte { return json64KPayload }

File: servers/hono/src/payload.ts
// Deterministic 1 KiB / 64 KiB JSON payload generator.
//
// Byte-identical port of probatorium/servers/common/payload.go. The Go
// reference uses encoding/json on (paginatedResponse, paginatedItem),
// which emits compact JSON with field order matching struct declaration
// order: page, per_page, total, total_pages, data for the wrapper, and
// id, name, email, status, created_at per item.
//
// We emit the bytes by hand here rather than via JSON.stringify on a
// JS object. Reasons:
//
//   1. Byte-for-byte equivalence with the Go reference is a hard
//      conformance requirement (cmd/conformance does bytes::Equal).
//      JSON.stringify and encoding/json agree on most things but offer
//      no formal cross-language guarantee, so we own the bytes.
//   2. The corpus is tiny and pure-ASCII (no escape hazards), so the
//      manual path is correct and trivially auditable.
//   3. Generated once at startup and reused for every request, so the
//      build cost is irrelevant.
//
// Termination rule from the Go reference: append items until the
// marshalled length is at least targetSize. Resulting sizes:
//   1 KiB target  -> 1026 bytes ending at item 9
//   64 KiB target -> 65618 bytes ending at item 583

const HEADER = '{"page":1,"per_page":50,"total":1000,"total_pages":20,"data":[';
const FOOTER = "]}";

let json1k: Uint8Array | undefined;
let json64k: Uint8Array | undefined;

export function json1KPayload(): Uint8Array {
  if (!json1k) json1k = generate(1024);
  return json1k;
}

export function json64KPayload(): Uint8Array {
  if (!json64k) json64k = generate(65536);
  return json64k;
}

function generate(targetSize: number): Uint8Array {
  // Emit straight into a string then encode once at the end. Bun's
  // string concatenation is rope-backed so per-iteration append is
  // cheap; we'd lose nothing by switching to a Bun.ArrayBufferSink, but
  // this runs exactly twice at process startup.
  let buf = HEADER;
  let i = 1;
  // Match the Go termination rule: stop once buf + footer length is at
  // least targetSize. The Go code does a full Marshal per iteration
  // which is equivalent — the wrapper struct serialises to a fixed
  // footer length, so length(buf) + length(footer) is exact.
  while (true) {
    if (i > 1) buf += ",";
    buf += item(i);
    if (buf.length + FOOTER.length >= targetSize) break;
    i += 1;
  }
  buf += FOOTER;
  // Pure-ASCII so byte length == char length; TextEncoder is correct
  // and idiomatic regardless.
  return new TextEncoder().encode(buf);
}

function item(n: number): string {
  // Order MUST match the Go struct declaration exactly:
  //   id, name, email, status, created_at.
  return (
    '{"id":' +
    n +
    ',"name":"User ' +
    n +
    '","email":"user' +
    n +
    '@example.com","status":"active","created_at":"2024-01-15T09:30:00Z"}'
  );
}
